### PR TITLE
Added Embedded Fonts config toggle

### PR DESCRIPTION
**Pull request by poire-z**
Just under the existing "Embedded Styles". Support for this was already there in crengine, and interfacing is similar to what is done with Embedded Styles.
Just pushing this so those interested can test it.
See the XXX, it may need a few other calls when untoggling Embedded Fonts, because the default font is often not used right after. But if one untoggle and retoggle Embedded Styles, it is then used. Same thing when changing fonts...
I haven't figured out yet a common pattern to make everything right, but we should :)

### CODE REVIEW

**Comment by poire-z** [comment]
OK, this problem of change of font not updating may be already there with Embedded Styles.
Could you confirm that this happens to you too (on a fresh koreader, without this patch):
- Launch a fresh koreader
- Open a .epub that has some styles, but NO embedded fonts.
- Change Font, pick a distinguishable font (but not a monospace) => the new font should be visible 
Toggle Embedded Styles multiple times : you should see the new font in all modes
- Change Font a 2nd time with another font =>the new font is visible only when Embedded Styles is Off, but the old one is visible when Embedded Style is On, and Embedded Style On will stick with this 1st changed font.

**Comment by Frenzie** [comment]
`Dracula - Bram Stoker.epub` from the test suite seems to have styles but no fonts. Maybe I missed something in your steps, but I don't think I can reproduce it. I tried it with embedded style both on and off before switching fonts again.

**Comment by poire-z** [comment]
OK, thanks. I checked with a Dracula epub with that name found on the web (marked feedbooks), and I, like you, do not reproduce the problem :)
So it must be something specific with the book i was testing with, which looked to me super simple in regards to embedded styles. Coud you check again with this book: [001description.epub.zip](https://github.com/koreader/koreader/files/1264029/001description.epub.zip)

Could be the presence of font-family:serif that i noticed before, that seem not there in the styles of this Dracula epub.

**Comment by Frenzie** [comment]
Yeah, I can confirm that. The behavior there is disorienting.

**Comment by poire-z** [comment]
Rebased with my XXX comments removed, as all strange things i had seen seem solved by the crengine PR referenced above.
There is no dependency between these 2 PRs, they can be merged in any order - it's just that without this crengine (and base+koreader) patches, people may see fonts not getting updated: the problem was already there with Embedded Styles, but is more noticable when playing with Embedded Fonts and changing default font to see how things look (before, we could accuse the embedded styles to be enforcing when we didnt see our new default font, that's what I thought :)

File: frontend/document/credocument.lua [hunk 1]
```diff
@@ -335,6 +335,26 @@ function CreDocument:setFontFace(new_font_face)
     if new_font_face then
         logger.dbg("CreDocument: set font face", new_font_face)
         self._document:setStringProperty("font.face.default", new_font_face)
+
+        -- The following makes FontManager prefers this font in its match
```
Comment (Frenzie) [inline frontend/document/credocument.lua]: meganitpick, but "prefer"

(who nitpicks comments? someone with an English degree, I suppose ;-P)

**Comment by j4w0r** [comment]
If there are no obstacles, can someone bump base with #504 and #505 and build new nightly?

**Comment by Frenzie** [comment]
I think this should be good to go, so I'll merge it.

**Comment by poire-z** [comment]
Thanks for the nightly! Working as expected on my kobo.

**Comment by j4w0r** [comment]
Frenzie thanks for merging and making new build. Poire-z, thank you very very much for bringing this feature to koreader. For now I have found one epub in which it does not work:

`<p class="Body">Some text with embeded font that can NOT be disabled</p>`

It has styles:
```
@font-face {
	font-family : "Liberation Serif";
	font-style : normal;
	font-weight : normal;
	src : url("../Fonts/LiberationSerif-Regular.otf");
}
body {
	font-family: "Liberation Serif", serif;
}
p.Body {
	font-family : "Liberation Serif", serif;
	font-weight : normal;
	font-style : normal;
	font-size : 1em;
	text-decoration : none;
	font-variant : normal;
	line-height : 1.2;
	text-align : justify;
	color : #000000;
	text-indent : 23px;
	margin : 0px;
}
```

There are styles that allow disabling embedded font, for example:
`<p class="BodySub">Some text with embeded font that can be disabled</p>`

```
@font-face {
	font-family : Alegreya;
	font-style : normal;
	font-weight : normal;
	src : url("../Fonts/Alegreya-Regular.otf");
}

p.BodySub {
	font-family : Alegreya, serif;
	font-weight : normal;
	font-style : normal;
	font-size : 1.17em;
	text-decoration : none;
	font-variant : normal;
	line-height : 1.2;
	text-align : left;
	color : #000000;
	text-indent : 0px;
	margin : 58px 0px 29px 0px;
}
```

**Comment by poire-z** [comment]
I tried to reproduce it by making some epub with the bits of code you gave: [005bodystyle.epub.zip](https://github.com/koreader/koreader/files/1272906/005bodystyle.epub.zip)
but it works, seen nothing strange.
Could you try it, and if it works for you, could you upload the original epub, so I may have a look ?

**Comment by j4w0r** [comment]
OK, it was the case from your "edit", I have had the Liberation Serif among koreader fonts. Is it bug or feature? :) In the meantime I have found another interesting case: [006bodystyle.epub.zip](https://ufile.io/j7znj). By the way, how can I upload exaple file like you did (I could not find it in repo)?

**Comment by poire-z** [comment]
> Is it bug or feature?

It's the way crengine's existing feature "Embedded Fonts On/Of" works:
One set of global fonts, one set of fonts found in the epub
When rendering the HTML, and looking for fonts to use according to each HTML node's style _font-family:_, it uses only the global set if "Embedded Fonts Off", and both sets if On. If font is not found, it uses our default font.
The HTML and styles associated to each node of the document (build at startup time) stays unchanged. So it's quite fast, and just some rendering work.

I would have prefered (i guess you too) a different way to do that:
When Embedded Fonts are Off, just ignore any _font-family:_ style (and just these) found in the embedded styles.
Alas, that would nearly need a whole re-opening and re-parsing of the document, and a lot more coding work (I did look at it, but gave up :)

But there is one option, if you really want only your default font (the only caveat i can think of is that monospace text would be lost and may look bad):
https://github.com/koreader/koreader/blob/dbaea8c1c8f49090b69251ddc3edb65992396adf/frontend/document/credocument.lua#L339-L357
You could change in that last line 1 to 256001, and see how it does for you.
That could be made into an option (with menu or not) if it indeed helps with some books.

(For uploading files, just drag and drop the file in that textarea you write in :)

**Comment by j4w0r** [comment]
Thank you for explanation and font tips - 256001 setting works good for me.

**Comment by robert00s** [comment]
256001 setting works great for me. I think its worth adding it as an option.

**Comment by Frenzie** [comment]
Where would we put it?

**Comment by robert00s** [comment]
What about something like that (force)?

obraz

**Comment by Frenzie** [comment]
"Force" means "force off", doesn't it? It might make more sense on the far right then.

**Comment by robert00s** [comment]
Right :)
obraz
or
obraz

**Comment by Frenzie** [comment]
Or maybe "never"? Reading @poire-z 's comment above I have to say I'm slightly confused by the behavior as it is.

**Comment by robert00s** [comment]
"never" sounds better

**Comment by robert00s** [comment]
Oh, 25601 is better option because with 256001 I cannot change to embedded (publisher) font...

**Pull request merged**